boa/src/builtins/reflect/tests.rs

```diff
@@ -0,0 +1,191 @@
+use crate::{forward, Context};
+
+#[test]
+fn apply() {
+    let mut context = Context::new();
+
+    let init = r#"
+        var called = {};
+        function f(n) { called.result = n };
+        Reflect.apply(f, undefined, [42]);
+        "#;
+
+    forward(&mut context, init);
+
+    assert_eq!(forward(&mut context, "called.result"), "42");
+}
+
+#[test]
+fn construct() {
```

Comment:
As an idea of something else to possibly also include:
```js
let x = Reflect.construct(String, ["test"], TypeError)
let e = new TypeError()
let res = Object.getPrototypeOf(x) == Object.getPrototypeOf(e)
// check that res is true
```
`x` will internally be `ObjectData::String`, but there isn't much in `boa` that makes it visible, might be possible in the future.

Surprisingly in `V8` the following code works ([this](https://tc39.es/ecma262/#sec-string-exotic-objects) seems to indicate that the string constructor should be adding certain properties):
> These methods are installed in StringCreate.

(from the spec)
```js
> for (a in x) { console.log(x[a])}
t
e
s
t
```

Reply:
This is a more general problem in `boa` with exotic objects, that is tracked in #602 for strings, and more generally in #591.
I tried to tackle it, but then I have the issue that all internal methods (as defined by the spec, that is `[[Get]]`, `[[Set]]` or `[[DefineOwnProperty]]` for instance) of a `GcObject` should have a parameter `&mut Context`, and if this is the case, then a `&mut Context` should also be added to nearly all methods of `EnvironmentTrait`, but then each time you change the environment is modified (when declaring a variable or calling a function), then you have to borrow `Context` mutably twice, once to get the `environment` and another time when you call a method of `EnvironmentTrait`. And this is of course not allowed by the Rust compiler.

@HalidOdat I saw your worked on this issue in PR #600, did you meet the same issue, or do I definitively miss something? I think this is important we can implement exotic objects properly, especially to support `Proxy`, typed arrays or even `String`.